### PR TITLE
DEPR: deprecate `astropy.utils.check_broadcast`

### LINKED ISSUE #15211: Deprecate check_broadcast
### Blocked by

* https://github.com/astropy/astropy/pull/15121

### Description

Numpy 1.20.0 added a function called [broadcast_shapes](https://numpy.org/doc/stable/reference/generated/numpy.broadcast_shapes.html#numpy.broadcast_shapes) which can replace check_broadcast from astropy.utils. Deprecate check_broadcast.

### Expected behavior

N/A

### How to Reproduce

N/A

### Versions

Astropy 5.3.2

### PR DESCRIPTION
**Pull request by neutrinoceros**
### Description
Fixes #15211 and replaces #15499 (I marked @felipekhouri as a co-author here)

I note that this implies `IncompatibleShapeError` is now unused (outside `check_broadcast`. Since users might be importing it to construct `try/except` blocks around `check_broadcast`, it should probably be deprecated too. Let me know if this should be done in the same PR or as a follow up.

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

File: astropy/modeling/core.py [hunk 1]
```diff
@@ -1063,11 +1065,16 @@ def _validate_input_shapes(self, inputs, argnames, model_set_axis):
...
+            input_shape = np.broadcast_shapes(*all_shapes)
+        except ValueError as exc:
+            note = "All inputs must have identical shapes or must be scalars."
+            if sys.version_info >= (3, 11):
+                exc.add_note(note)
+            else:
+                # mimic Python 3.11 behavior:
+                # preserve error message and traceback
+                exc = type(exc)(f"{exc!s}\n{note}").with_traceback(exc.__traceback__)
+            raise exc
```
Comment (neutrinoceros) [inline astropy/modeling/core.py:1077]: This is a side improvement but isn't strictly needed here, I just applied an existing snippet of mine here. Let me know if I should revert it or move it to a separate PR.

File: astropy/modeling/core.py [hunk 2]
```diff
@@ -2229,8 +2236,8 @@ def _prepare_outputs_single_model(self, outputs, broadcasted_shapes):
         outputs = list(outputs)
         for idx, output in enumerate(outputs):
             try:
-                broadcast_shape = check_broadcast(*broadcasted_shapes[0])
-            except (IndexError, TypeError):
+                broadcast_shape = np.broadcast_shapes(*broadcasted_shapes[0])
+            except Exception:
```
Comment (neutrinoceros) [inline astropy/modeling/core.py:2240]: note that I made the error catching *less* accurate here so it would also capture `DeprecationWarning`s when they're elevated to errors in testing, and which should become actual exceptions of unknown types in the future.

File: astropy/modeling/core.py [hunk 3]
```diff
@@ -2742,18 +2749,40 @@ def _check_param_broadcast(self, max_ndim):
...
                 "to the broadcasting rules."
             )
+            broadcast_shapes_error_re = re.compile(
+                r"shape mismatch: objects cannot be broadcast to a single shape\.  "
+                r"Mismatch is between "
+                r"arg (?P<argno_a>\d+) with shape (?P<shape_a>\((\d+(, ?)?)+\)) and "
+                r"arg (?P<argno_b>\d+) with shape (?P<shape_b>\((\d+(, ?)?)+\))\."
+            )
+            if (match := broadcast_shapes_error_re.fullmatch(str(exc))) is None:
+                warnings.warn(
```
Comment (neutrinoceros) [inline astropy/modeling/core.py:2766]: Since parsing the error message emitted by numpy is intrinsically fragile (but probably [our least-bad option](https://github.com/astropy/astropy/issues/15211#issuecomment-2076613999)), I spent extra effort on making this logic as robust as possible: failing to parse an error message should be recoverable.

File: astropy/modeling/core.py [hunk 4]
```diff
@@ -1949,7 +1956,7 @@ def _prepare_inputs_single_model(self, params, inputs, **kwargs):
             for param in params:
                 try:
                     if self.standard_broadcasting:
-                        broadcast = check_broadcast(input_shape, param.shape)
+                        broadcast = np.broadcast_shapes(input_shape, param.shape)
                     else:
                         broadcast = input_shape
                 except IncompatibleShapeError:
```
Comment (neutrinoceros) [inline astropy/modeling/core.py:1962]: Whoops, I should have updated this line too.

File: astropy/utils/exceptions.py [hunk 5]
```diff
@@ -75,6 +75,18 @@ def __repr__(self):
...
+def _add_note_to_exception(exc: Exception, note: str) -> Exception:
+    import sys
+
+    if sys.version_info >= (3, 11):
+        exc.add_note(note)
+        return exc
+    else:
+        # mimic Python 3.11 behavior:
+        # preserve error message and traceback
+        return type(exc)(f"{exc!s}\n{note}").with_traceback(exc.__traceback__)
```
Comment (neutrinoceros) [inline astropy/utils/exceptions.py:87]: This is a side improvement but isn't strictly needed here, I just applied an existing snippet of mine here. Let me know if I should revert it or move it to a separate PR.

**Review by mhvk — COMMENTED** [review]
Thanks, this looks mostly excellent. But I really would prefer no code changes/comments just for test behaviour - hence the query whether that could be dealt with in the tests themselves. The rest is mostly small stuff.

File: astropy/utils/exceptions.py (hunk 5 above)
Comment (neutrinoceros) [reply astropy/utils/exceptions.py:87]: Hum, I didn't think about that. I guess replacing a tuple attribute is objectively less evil than returning an entirely new instance in a context where we accept mutations anyway. Thanks !

File: astropy/modeling/core.py [hunk 6]
```diff
@@ -2229,8 +2235,12 @@ def _prepare_outputs_single_model(self, outputs, broadcasted_shapes):
         outputs = list(outputs)
         for idx, output in enumerate(outputs):
             try:
-                broadcast_shape = check_broadcast(*broadcasted_shapes[0])
-            except (IndexError, TypeError):
+                # always promote deprecation warnings to exceptions
```
[the remark this answers is not in the document]
Comment (neutrinoceros) [reply astropy/modeling/core.py:2238]: > Is this really necessary?

For reference, the one test that failed before I applied this change was `modeling/tests/test_core.py::test_inputless_model`; It failed *only in the context of wheel testing* where the relevant difference with regular CI is that warnings are not elevated to errors.
See [example logs](https://github.com/astropy/astropy/actions/runs/8844042785/job/24285246022)
I haven't found a way to adjust the test itself yet, but I agree that this patch leaves a bad taste. I'll circle back to this.

> p.s. Wish I could trust that except Exception was not covered by the tests...

From my experience running this is CI, it *is* covered, or I wouldn't have needed the aforementioned patch.

File: astropy/utils/tests/test_shapes.py [hunk 7]
```diff
@@ -1,22 +1,32 @@
...
+from astropy.utils.exceptions import AstropyDeprecationWarning
 from astropy.utils.shapes import check_broadcast, simplify_basic_index, unbroadcast
 
 
+def test_check_broadcast_deprecation():
+    with pytest.warns(AstropyDeprecationWarning):
+        check_broadcast((1,), (2,))
+
+
 def test_check_broadcast():
```
[the remark this answers is not in the document]
Comment (neutrinoceros) [reply astropy/utils/tests/test_shapes.py:20]: That would have been my preference too, unfortunately this doesn't seem to work properly with non-builtin exception types, and causes pytest to crash on collection.

**Comment by neutrinoceros** [comment]
@mhvk thanks for your review. I think I addressed all your minor comments. I'll circle back to [the more important one](https://github.com/astropy/astropy/pull/16346#discussion_r1581833541) on Monday !

File: astropy/coordinates/tests/test_frames.py [hunk 8]
```diff
@@ -259,7 +260,11 @@ def test_no_data_nonscalar_frames():
             obstime=Time("2012-01-01") + np.arange(10.0) * u.day,
             temperature=np.ones((3,)) * u.deg_C,
         )
-    assert "inconsistent shapes" in str(exc.value)
```
Comment (mhvk) [inline astropy/coordinates/tests/test_frames.py:262]: Actually, looking at this again, a question: I assume the "inconsistent shapes" appear in the text the user sees, since otherwise it is not very useful. If so, why isn't it in `str(exc)`? Would it be in `repr(exc)`? Ideally, we test as closely as possible what the user sees, not how it is implemented that they get to see it.

File: astropy/utils/tests/test_shapes.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (mhvk) [reply astropy/utils/tests/test_shapes.py:20]: Maybe just ignore all warnings? But no big deal, obviously.

File: astropy/coordinates/tests/test_frames.py (hunk 8 above)
Comment (neutrinoceros) [reply astropy/coordinates/tests/test_frames.py:262]: I learned this while updating this test but notes are apparently not included in `__str__` or `__repr__`, but this is an implementation detail not visible by a user when an exception is raised.

Comment (mhvk) [reply astropy/coordinates/tests/test_frames.py:262]: Ah, so the user sees it, but `str` or `repr` doesn't show it. That seems like a bug in python to me, or arguably `pytest`, which should really just match what the user sees. Anyway, for now this will have to do!

Comment (neutrinoceros) [reply astropy/coordinates/tests/test_frames.py:262]: I was curious so I looked it up and I'm not convinced it can be called a bug given [that `str()` is documented to represent instance arguments](https://docs.python.org/3/library/exceptions.html#BaseException). A shame that it forces us to test implantation details but I guess that's the price to pay for backward compatibility.

Comment (mhvk) [reply astropy/coordinates/tests/test_frames.py:262]: Yes, I was just looking up things myself too, and came to the same conclusion. So, not a python bug, nor a pytest bug if one takes the description of `match` literally, though I think it might sense for `pytest` to support matching on anything the user would see. 
...
But then, of course I found that others had thought the same way, so newer pytest already includes `__notes__` in what it matches: https://github.com/pytest-dev/pytest/pull/11227. I think this has entered pytest 8.0 (the first release after the merge was 7.4.1, but seems unlikely it was included in a bugfix release).

In the meantime, maybe add a comment and link this PR in #15180?

Comment (neutrinoceros) [reply astropy/coordinates/tests/test_frames.py:262]: Excellent, thank you !

File: astropy/modeling/core.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (neutrinoceros) [reply astropy/modeling/core.py:2238]: note to self: I think `Model._prepare_inputs_single_model` underwent an implicit change of behavior and needs adjusting.

[the remark this answers is not in the document]
Comment (neutrinoceros) [reply astropy/modeling/core.py]: I just pushed what I think is a more maintainable solution, together with a companion comment explaining my findings.

**Review by mhvk — APPROVED** [review]
Looks good! 

Final question: did you try ignoring all warnings in `test_check_broadcast`? If that also doesn't work, then let's get things in as they are. Thanks!

**Comment by neutrinoceros** [comment]
Thanks for reminding me. I updated the test with your suggestion then squashed all commits together.
